crypto/merklearray/proof.go

```diff
@@ -16,36 +16,51 @@
 
 package merklearray
 
-import "github.com/algorand/go-algorand/crypto"
+import (
+	"github.com/algorand/go-algorand/crypto"
+)
 
-// Proof contains the merkle path, along with the hash factory that should be used.
+// Proof is used to convince a verifier about membership of leaves: h0,h1...hn
+// at indexes i0,i1...in on a tree. The verifier has a trusted value of the tree
+// root hash.
 type Proof struct {
 	_struct struct{} `codec:",omitempty,omitemptyarray"`
 
-	// Path is bounded by MaxNumLeaves since there could be multiple reveals, and
+	// Path is bounded by MaxNumLeavesOnEncodedTree since there could be multiple reveals, and
 	// given the distribution of the elt positions and the depth of the tree,
-	// the path length can increase up to 2^MaxTreeDepth / 2
-	Path        []crypto.GenericDigest `codec:"pth,allocbound=MaxNumLeaves/2"`
+	// the path length can increase up to 2^MaxEncodedTreeDepth / 2
+	Path        []crypto.GenericDigest `codec:"pth,allocbound=MaxNumLeavesOnEncodedTree/2"`
 	HashFactory crypto.HashFactory     `codec:"hsh"`
+	// TreeDepth represents the depth of the tree that is being proven.
+	// It is the number of edges from the root to a leaf.
+	TreeDepth uint8 `codec:"td"`
 }
 
-// GetSerializedProof serializes the proof into a sequence of bytes.
-// it basically concatenates all the verification path one after another.
-// The function returns a fixed length array for each hash function. which is 1 + MaxTreeDepth * digestsize
+// SingleLeafProof is used to convince a verifier about membership of a specific
+// leaf h at index i on a tree. The verifier has a trusted value of the tree
+// root hash. it corresponds to merkle verification path.
+type SingleLeafProof struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+	Proof
+}
+
+// GetFixedLengthHashableRepresentation serializes the proof into a sequence of bytes.
+// it basically concatenates the elements of the verification path one after another.
+// The function returns a fixed length array for each hash function. which is 1 + MaxEncodedTreeDepth * digestsize
 //
-// the path is guaranteed to be less than MaxTreeDepth and if the path length is less
-// than MaxTreeDepth, array is padded with zeros.
+// the path is guaranteed to be less than MaxEncodedTreeDepth and if the path length is less
+// than MaxEncodedTreeDepth, array is padded with zeros.
 // more details could be found in the Algorand's spec.
-func (p *Proof) GetSerializedProof() []byte {
+func (p *SingleLeafProof) GetFixedLengthHashableRepresentation() []byte {
 	hash := p.HashFactory.NewHash()
 
-	var binProof = make([]byte, 0, 1+(MaxTreeDepth*hash.Size()))
+	var binProof = make([]byte, 0, 1+(MaxEncodedTreeDepth*hash.Size()))
```

Comment:
Should there be an error in case the tree size is larger than `MaxEncodedTreeDepth`?